File: Swift-C++-Testsuite/test/protocols/Swift/ProtocolsAsTypes.swift
@main
struct MyMain {

    static func main() {
        var d6 = Dice(sides: 6, generator: LinearCongruentialGenerator())
        for _ in 1...5 {
            print("Random dice roll is \(d6.roll())")
        }
    }
}

class LinearCongruentialGenerator: RandomNumberGenerator {
    var lastRandom = 42.0
    let m = 139968.0
    let a = 3877.0
    let c = 29573.0
    func random() -> Double {
        lastRandom =
            ((lastRandom * a + c)
                .truncatingRemainder(dividingBy: m))
        return lastRandom / m
    }
}

class Dice {
    let sides: Int
    let generator: RandomNumberGenerator
    init(sides: Int, generator: RandomNumberGenerator) {
        self.sides = sides
        self.generator = generator
    }
    func roll() -> Int {
        return Int(generator.random() * Double(sides)) + 1
    }
}

protocol RandomNumberGenerator {
    func random() -> Double
}

File: Swift-C++-Testsuite/test/protocols/C++/ProtocolsAsTypes.cpp
 
 #include <iostream>
 
 class RandomNumberGenerator
 {
 	public:
 	virtual double random() = 0;
 };
 
 class LinearCongruentialGenerator : public RandomNumberGenerator
 {
 	public:
 	double lastRandom = 42.0;
 	const double m = 139968.0;
 	const double a = 3877.0;
 	const double c = 29573.0;
 	
 	virtual double random()
 	{
 		lastRandom = ((lastRandom * a + c) / m);
 		return lastRandom;
 	}
 
 };
 
 class Dice
 {
 	public:
 	int sides;
 	RandomNumberGenerator *generator;
 	Dice(int sides, RandomNumberGenerator *gen)
 		: sides(sides), generator(gen) { }
 	int roll()
 	{
 		double r = generator->random();
 		return int(r * sides + 1);
 	}
 };
 
 int main()
 {
 	LinearCongruentialGenerator lcg;
 	Dice d6(6, &lcg);
 	for(int i = 0; i < 5; ++i)
 		std::cout << "Random dice roll is " << d6.roll() << std::endl;
 		
 	return 0;
 }